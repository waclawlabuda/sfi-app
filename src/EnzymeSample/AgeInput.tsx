import React from 'react';

export interface EnzymeSampleProps {
  onAgeChange(e: any): void;
  age: string;
}

class AgeInput extends React.Component<EnzymeSampleProps, {}> {

  onChange = (e) => {
    this.props.onAgeChange(e);
  };

  render() {
    const { age } = this.props;

    return (
      <input type="number" id="age_input" placeholder="Age" value={age} onChange={this.onChange} />
    );
  }
}

export default AgeInput;
