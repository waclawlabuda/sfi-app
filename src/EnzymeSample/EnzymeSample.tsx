import React from 'react';

export interface EnzymeSampleState {
   name: string;
   age: string;
   checked: boolean;
   error: boolean;
}

export interface EnzymeSampleProps {
   className?: string;
   onSubmit(data?: Pick<EnzymeSampleState, 'name' | 'age' | 'checked'>): void;
}

class EnzymeSample extends React.Component<EnzymeSampleProps, EnzymeSampleState> {
   state = {
      name: '',
      age: '',
      checked: false,
      error: false,
   }

   onNameChange = (e) => {
      const newName = e.target.value;
      if (/\d/.test(newName)) {
         this.setState({ error: true });
      } else {
         this.setState({ error: false });
      }
      this.setState({ name: e.target.value });
   }

   onAgeChange = (e) => {
      this.setState({ age: e.target.value });
   }

   onCheckboxChange = () => {
      this.setState((prevState) => ({ checked: !prevState.checked }));
   }

   onSubmitHandle = () => {
      this.props.onSubmit({
         name: this.state.name,
         age: this.state.age,
         checked: this.state.checked,
      });

      this.setState({
         name: '',
         age: '',
         checked: false,
      });
   }

   render() {
      const { name, age, checked, error } = this.state;
      const { className } = this.props;

      return (
         <div className={className}>
            <h2>Some random heading</h2>
            <form className='form'>
               <div>
                  {error && <span className='error'>name cannot have numbers</span>}
                  <br />
                  <input
                     type='text'
                     id="name_input"
                     placeholder='Write your name'
                     value={name}
                     onChange={this.onNameChange}
                  />
                  <br />
                  <input
                     type='number'
                     id="age_input"
                     placeholder='Age'
                     value={age}
                     onChange={this.onAgeChange}
                  />
                  <br />
                  <br />
                  <label
                     htmlFor='checkbox'
                  >
                     <input
                        id='checkbox'
                        type='checkbox'
                        checked={checked}
                        onChange={this.onCheckboxChange}
                     />
                     check me!
                  </label>
               </div>
               <button onClick={this.onSubmitHandle} disabled={!Boolean(name && age && checked && !error)}>Submit!</button>
            </form>
         </div>
      )
   }
}

export default EnzymeSample;