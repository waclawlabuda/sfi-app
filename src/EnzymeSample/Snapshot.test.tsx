import React from 'react';
import testRenderer from 'react-test-renderer';
import EnzymeSample, { EnzymeSampleProps, EnzymeSampleState } from './EnzymeSample';
import { ShallowWrapper, ReactWrapper, mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

const setup = (propsOverride?: Partial<EnzymeSampleProps>) => {
    const props = {
        onSubmit: jest.fn(),
        ...propsOverride,
    }

    const renderer = () => testRenderer.create(
        <EnzymeSample {...props} />
    );

    const component: ShallowWrapper<EnzymeSampleProps, EnzymeSampleState> = shallow(<EnzymeSample {...props} />);
    const mountedComponent: ReactWrapper = mount(<EnzymeSample {...props} />);

    return {
        props,
        renderer,
        component,
        mountedComponent,
    }
}

describe('EnzymeSample', () => {
    it('should render properly', () => {
        const { renderer } = setup();
        expect(renderer().toJSON()).toMatchSnapshot();
    });

    it('should render properly', () => {
        const { mountedComponent } = setup();
        expect(toJson(mountedComponent)).toMatchSnapshot();
    });

    it('should display error message when user type number into name input', () => {
        const { component } = setup();

        expect(component.state('error')).toBeFalsy();
        expect(component.find('.error')).toHaveLength(0);

        component.find('#name_input').simulate('change', { target: {
            value: '1234'
        }});

        expect(component.state('error')).toBeTruthy();
        expect(component.find('.error')).toHaveLength(1);
    });

    it('should change state on checkbox click (checked -> true)', () => {
        const { component } = setup();

        expect(component.state('checked')).toBeFalsy();

        component
            .find('label')
            .find('input')
            .simulate('change');

        expect(component.state('checked')).toBeTruthy();
    });

    it('should be disabled if inputs are invalid', () => {
        const { component } = setup();
        const button = component.find('button');
        expect(button.prop('disabled')).toBeTruthy();
    });

    it('should be enabled if inputs are valid', () => {
        const { component } = setup();
        component.setState({
            name: 'John Doe',
            age: '44',
            checked: true,
            error: false,
        });
        const button = component.find('button');
        expect(button.prop('disabled')).toBeFalsy();
    });

    it('should call onSubmit when button was clicked', () => {
        const { component, props } = setup();
        component.find('button').simulate('click');
        expect(props.onSubmit).toBeCalled();
    });

    it('should clear all fields on button click', () => {
        const { component } = setup();
        component.setState({
            name: 'John Doe',
            age: '44',
            checked: true,
        });

        const button = component.find('button');
        button.simulate('click');

        expect(component.state('name')).toBe('');
        expect(component.state('age')).toBe('');
        expect(component.state('checked')).toBeFalsy();
    });
});