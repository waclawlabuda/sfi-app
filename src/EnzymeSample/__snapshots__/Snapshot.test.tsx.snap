// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnzymeSample should render properly 1`] = `
<div>
  <h2>
    Some random heading
  </h2>
  <form
    className="form"
  >
    <div>
      <br />
      <input
        id="name_input"
        onChange={[Function]}
        placeholder="Write your name"
        type="text"
        value=""
      />
      <br />
      <input
        id="age_input"
        onChange={[Function]}
        placeholder="Age"
        type="number"
        value=""
      />
      <br />
      <br />
      <label
        htmlFor="checkbox"
      >
        <input
          checked={false}
          id="checkbox"
          onChange={[Function]}
          type="checkbox"
        />
        check me!
      </label>
    </div>
    <button
      disabled={true}
      onClick={[Function]}
    >
      Submit!
    </button>
  </form>
</div>
`;

exports[`EnzymeSample should render properly 2`] = `
<EnzymeSample
  onSubmit={[MockFunction]}
>
  <div>
    <h2>
      Some random heading
    </h2>
    <form
      className="form"
    >
      <div>
        <br />
        <input
          id="name_input"
          onChange={[Function]}
          placeholder="Write your name"
          type="text"
          value=""
        />
        <br />
        <AgeInput
          age=""
          onAgeChange={[Function]}
        >
          <input
            id="age_input"
            onChange={[Function]}
            placeholder="Age"
            type="number"
            value=""
          />
        </AgeInput>
        <br />
        <br />
        <label
          htmlFor="checkbox"
        >
          <input
            checked={false}
            id="checkbox"
            onChange={[Function]}
            type="checkbox"
          />
          check me!
        </label>
      </div>
      <button
        disabled={true}
        onClick={[Function]}
      >
        Submit!
      </button>
    </form>
  </div>
</EnzymeSample>
`;
